qualtran/resource_counting/_success_prob.py

```diff
@@ -0,0 +1,49 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import logging
+from typing import Callable
+
+from attrs import frozen
+
+from qualtran import Bloq
+
+from ._call_graph import get_bloq_callee_counts
+from ._costing import CostKey
+
+logger = logging.getLogger(__name__)
+
+
+@frozen
+class SuccessProb(CostKey[float]):
+    """The success probability of a bloq.
+
+    A bloq's success probability is the multiplicative product of its callees'
+    success probabilities. Bloqs that have a specific success probability should override
+    `my_static_costs` to provide their actual success probability.
+    """
+
+    def compute(self, bloq: 'Bloq', get_callee_cost: Callable[['Bloq'], float]) -> float:
+        tot: float = 1.0
+        callees = get_bloq_callee_counts(bloq)
+        logger.info("Computing %s for %s from %d callee(s)", self, bloq, len(callees))
+        for callee, n in callees:
+            v = get_callee_cost(callee)
+            tot *= v**n
```

Comment:
The way this is written, would we ever need this as a separate cost? Can we just use `g, sigma = bloq.build_call_graph(pred=pred)` and then accumulate the error probability using bloq, counts in `sigma` ? 

Is there a good example that motivates this cost key that already exists in another branch / is easy to describe ?

Reply:
you need the contributions from inner nodes as well, not just the leafs that will show up in `sigma`. 

This was requested by someone at the IEEE meeting last year and is a good demo of a different type of cost